### PR TITLE
Use Resource labels as additional arguments for Zipkin. Also re-use the service name

### PR DESCRIPTION
**Pull request by SergeyKanzhelev**

### CODE REVIEW

**Comment by SergeyKanzhelev** [comment]
@open-telemetry/dotnet-approvers can you please take a look?

File: src/OpenTelemetry/Trace/Configuration/TracerBuilder.cs [hunk 1]
```diff
@@ -50,6 +53,29 @@ public TracerBuilder SetSampler(ISampler sampler)
...
+            this.Resource = resource ?? Resource.Empty;
+            return this;
+        }
+
+        /// <summary>
+        /// Merge the provided <see cref="Resource"/> with one currently set by builder.
+        /// </summary>
+        /// <param name="resource">Resource to associate with all traces.</param>
+        /// <returns>Trace builder for chaining.</returns>
+        public TracerBuilder MergeResource(Resource resource)
```
Comment (lmolkova) [inline src/OpenTelemetry/Trace/Configuration/TracerBuilder.cs]: why do we need this method? Can user merge all necessary resources and then call SetResource with final one?

File: README.md [hunk 2]
```diff
@@ -262,11 +263,12 @@ Configuration is done by user application: it should configure exporter and may
...
         builder
             .SetSampler(Samplers.AlwaysSample)
-            .UseZipkin(o => o.ServiceName = "my-service")
+            .UseZipkin()
 
             // you may also configure request and dependencies collectors
             .AddRequestCollector()
-            .AddDependencyCollector())
+            .AddDependencyCollector()
+            .MergeResource(new Resource(new Dictionary<string, string>() { { "service.name", "my-service" } }))
```
Comment (lmolkova) [inline README.md]:
```suggestion
            .SetResource(new Resource(new Dictionary<string, string>() { { "service.name", "my-service" } }))
```

File: README.md [hunk 3]
```diff
@@ -262,11 +263,12 @@ Configuration is done by user application: it should configure exporter and may
     {
         builder
             .SetSampler(Samplers.AlwaysSample)
-            .UseZipkin(o => o.ServiceName = "my-service")
```
Comment (lmolkova) [inline README.md]: should we remove ServiceName from zipkin (and possibly other exporters) options?

File: src/OpenTelemetry/Trace/Configuration/TracerBuilder.cs [hunk 4]
```diff
@@ -50,6 +53,29 @@ public TracerBuilder SetSampler(ISampler sampler)
             return this;
         }
 
+        /// <summary>
+        /// Sets the <see cref="Resource"/> describing the app associated with all traces. Overwrites currently set resource.
```
Comment (lmolkova) [inline src/OpenTelemetry/Trace/Configuration/TracerBuilder.cs]: TracerBuilder is single-use thing. It is not intended for any update scenarios. Basically you use it to construct factory and throw it away.

Runtime changes are not possible at all. So Set always overwrites and Merge is overkill :)

Comment (SergeyKanzhelev) [reply src/OpenTelemetry/Trace/Configuration/TracerBuilder.cs]: I should have explained better. `Set` and `Merge` are different in semantics and needed. THe idea is that there are multiple resource collectors:
- One that reads environment variables
- one that set by user explicitly
- one that understands specific environment (ex Azure Web Apps)
etc.

One that will be pre-installed is that reads environment variables. User may choose to omit it if using `Set`.

File: README.md (hunk 3 above)
Comment (SergeyKanzhelev) [reply README.md]: I didn't remove it to allow to set up the "default" service name. I agree though I'd rather switch to resource model all together

File: README.md (hunk 2 above)
Comment (SergeyKanzhelev) [reply README.md]: Nope. User intent here is to add couple properties, not to override the default one.

File: src/OpenTelemetry/Trace/Configuration/TracerBuilder.cs (hunk 1 above)
Comment (SergeyKanzhelev) [reply src/OpenTelemetry/Trace/Configuration/TracerBuilder.cs]: explained in different comment. I'll put more details in comments for the method

File: src/OpenTelemetry/Trace/Configuration/TracerBuilder.cs (hunk 4 above)
Comment (lmolkova) [reply src/OpenTelemetry/Trace/Configuration/TracerBuilder.cs]: Something can merge all resources into one and set a default one on the TracerBuilder. Why does TracerBuilder needs to provide this capabilities?

Comment (SergeyKanzhelev) [reply src/OpenTelemetry/Trace/Configuration/TracerBuilder.cs]: - I want to initialize TraceBuilder with the default one. And have an ability to override it. 
- Also I'd love to make calculation of resource lazy.
- and finally I can imagine some helper methods like AddAzureSupport() and AddK8SSupport() in TracerBuilder that adds whole bunch of resource and other things.

Thus I need some sort of resopurces pipeline builder. Alternative for this is to create a resources pipeline builder as an explicit type and set it on TracerBuilder. Are you advocating for this?

Comment (lmolkova) [reply src/OpenTelemetry/Trace/Configuration/TracerBuilder.cs]: Service name should be part of default resource, isn't it?
You should not reuse the same builder twice, so you would anyway construct one resource per builder.

AddAzureSupport or AddK8Ssupport are future and we are not designing them now. So how they will work - let decide later. Now it seems the scenario you want to have is to assign resource to the factory. So why can't we have one method for it and introduce complications only when they are needed.

Comment (MikeGoldsmith) [reply src/OpenTelemetry/Trace/Configuration/TracerBuilder.cs]: I like having both Set and Merge methods. It's clear they have different functions and it's nice to be able to replace the whole resource, or merge entries.

**Comment by SergeyKanzhelev** [comment]
We can discuss Set/Merge needs later when they will emerge

**Pull request merged**